### DISCUSSION TITLE
Scraper management system design

### DISCUSSION THREAD

**Discussion by WithoutPants** [stashapp/stash#2244] [Ideas]
For the implementation fo #623 

# Identified use cases

* edit scraper repositories
* install scrapers from repository
* view out of date scrapers and update
* remove scrapers

# Design Decisions

* it should be possible to set the source of scrapers to download
* the scraper repository needs an index file to list scraper details
* local scrapers will be matched to repository scrapers by filename
* the community scrapers repo should use github pages to serve the static versions of the scraper index file and scraper files
* scraper management operations should use the task subsystem
* TODO - checksum or update date to determine need to update
* it should be possible to view and remove scrapers not part of a repository

## Scraper sources

The community scraper repository is currently the primary (and arguably only) source of scrapers, but I do not want to hard code a link into stash. We want to allow for alternative repositories. I think that we should include the official repo by default, but we should allow the configuration of additional or replacement repositories.

## Scraper index file

In order to show useful information about scrapers available to download or update, we need to get metadata about each scraper. Ideally, we should be able to get this information without downloading the actual scraper file itself.

I propose that we have a generated yml index file that lists the location and details about each available scraper. This way, stash can download this index file and provide a useful list of scrapers, along with a description, capabilities and supported URLs. It should also provide information to determine if a scraper needs updating.

Possible alternatives:
* only list the files without any details - doesn't give the user much information
* download the entire repository and read the files themselves. This downloads too much information for a simple check.

## Scraper matching

A local scraper will be considered a match for a repository scraper if the filename is the same. I don't think there's any reasonable alternative to this.

## Github pages

We need the ability to download the scraper files (and the index file). These files are currently stored in a github repository. The only way to download these files is to clone the repository, download the zip, or view the raw versions of the files. As stated [here](https://github.blog/2013-04-24-heads-up-nosniff-header-support-coming-to-chrome-and-firefox/), linking directly to the raw renders of the files causes github server load, and is not good practice for us.

I propose instead that we automate the generation of the index file and deploy all files statically using github pages.

## Scraper operations

While it should be possible to install, update and remove scrapers individually, it is likely that users will install and update many scrapers at once. Therefore, I think that the operations should be performed using the task subsystem. 

There are potential issues with this approach. Ideally, scraper jobs should be started immediately and run concurrently with any other running tasks. We don't want to have the user waiting to install or update scrapers because they are currently scanning or generating. To that end, we will also want to ensure that only one scraper operation can occur at once. 

## Determining when a scraper needs updating

There needs to be some way for stash to determine if a scraper needs to be updated.

One approach is to using scraper file checksums. Using scraper file checksum means that stash would generate the file checksum of each scraper file and compare it against the checksum in the repository (listed in the index file). The downside to this approach is that it only indicates that the files are different, not that they are out of date.

Another approach is to use commit date of the scraper files. This could be calculated during the index file generation process. Stash would need some way to determine the date of the current scraper file. One approach is to manually set the modification time on the file and use that to compare. This approach will come unstuck if the user subsequently touches the file. Another approach is to write the modification time into the file during the install/update process. This means that the file is the different from the source file.

## Management of scrapers outside repositories

Scrapers can be manually created and sourced from places outside of repositories. We want the maintain the ability to view these scrapers and allow removing the scrapers via the UI.

# Use cases and UI

## Editing scraper repositories

Scraper repositories should be listed under `Metadata Providers` in the same way as stash-box endpoints. Scraper repositories will have a name and URL field. 

## Installing and updating scrapers

* user goes to Settings -> Metadata Providers
* under the Scraper Repositories section is a list of scraper repositories. Under each repository entry, a "Install/Update" button will be shown.
* user clicks on the "Install/Update" button for the repository they want to install/update scrapers from.
* system downloads and processes the index file for the repository.
* system displays a dialog with a table of scrapers to update and install. This includes filename, description, scraping capabilities and supported URLs. Each entry will have buttons to mark the scraper as to (re-)install, update or remove.

The UI will include options to filter by text (name, description and URLs). It will also have the ability to show only installed scrapers, not installed scrapers, or out of date scrapers. The UI should allow for marking all to install/update. The UI should mark all out of date scrapers to update by default.

* user clicks buttons to install missing scrapers, update out of date scrapers, and remove unwanted scrapers
* user clicks the Apply button
* system starts a task to apply the scraper changes, and indicates the running operation in the UI.

## Removing scrapers not managed by repositories

The Scrapers list will be changed to add a button to remove a scraper. A confirmation dialog will be shown before the scraper is removed. This will be an immediate operation and will not use the task subsystem.

**Comment by kermieisinthehouse** [comment]
This will necessitate a change to the scraper format, but scrapers should declare the resources they need. For example, a popular scraper needs an API key to function. The scraper should declare that it needs a string passed into it and this UI should require the user to fill it in before making the scraper available. If a scraper is providing a "manifest" already, it should perhaps also declare what URLs it might access, so that we can inform the user of the privacy implications of a scraper.

Also, how might we handle dependencies? I'm not completely caught up on scene scrapers (I only use basic performer scrapers), but I understand a few require python and some libs installed. Should we have python installation instructions and automate some `pip install`s? Should scrapers declare their dependencies too? I also believe there's a common python class that some scrapers share, not sure on the details.

**Comment by bnkai** [comment]
Design Decisions
- As mentioned above a lot of scrapers have some configurable sections. The first thing we need to decide imho is how to deal with it. We want to separate the configuration from the actual scraper so that we can then check what to update.
XPath or JSON scrapers should have a way to separate the configuration either in a separate file or as mentioned some time ago in a separate config section of stash.
Script scrapers can deal with that probably with a separate file as most existing ones already do (some in separate files, some share a common file at least for the stash specific options )
This will probably need a separate RFC as there are some complications and implementation details that will need to be checked with more detail
Depending on the decision for the above the  UI could either check for config options or just open a text editor with the configuration file

- A `manifest` file  is probably a good way to have all scraper capabilites/dependencies( as in local files needed by scripts, configuration files)/info/changelog/usage details for non url scrapers and anything else that is needed. A `generator` could then parse the file and create an index for all the scrapers

- **local scrapers will be matched to repository scrapers by filename**
An alternative would be the name of the scraper as mentioned in the yaml file. For example i rename the actual yaml files to alter the order of the scrapers in the dropdown list. The name still remains the same though (and it should be available in the index file)

- **Determining when a scraper needs updating**
The current approach relies on the fact that every scraper (xpath/json/script) has a `yml` file
The validator by @peolic  [validator](https://github.com/stashapp/CommunityScrapers#validation) is run to ensure that every yaml file has a  last updated comment with a specific format so that the existing plugin by @Belleyy [GHScraperChecker](https://github.com/stashapp/CommunityScripts/tree/main/plugins/GHScraper_Checker) can check that out.  Any change in a scraper file (especially for scripts) requires(we double check that manually atm before merging the PR) that the commented date in the yml file is updated
This is probably a good starting point as the data is already in the yml file and could be parsed or propagated to the manifest or the index file

- **Scraper operations**
Scraping operation should be run separately from the tasks. It is very common to do some scraping while waiting for scan/generation to finish. That would mean though that the identify task should be also moved? Allowing a single scrape at a  time will also  make it easier to limit/throttle requests during an identify task.

- **Github pages**
I was not aware of the issue. If the index/manifest file is used i think generating static pages would indeed be way to go

**Reply by WithoutPants** [reply]
> * As mentioned above a lot of scrapers have some configurable sections. The first thing we need to decide imho is how to deal with it. We want to separate the configuration from the actual scraper so that we can then check what to update.

I'm working on a scraper parameter PR currently as a prerequisite for this.

>   An alternative would be the name of the scraper as mentioned in the yaml file. For example i rename the actual yaml files to alter the order of the scrapers in the dropdown list. The name still remains the same though (and it should be available in the index file)

This sounds reasonable. I will amend the design.

> * **Determining when a scraper needs updating**
>   The current approach relies on the fact that every scraper (xpath/json/script) has a `yml` file
>   The validator by @peolic  [validator](https://github.com/stashapp/CommunityScrapers#validation) is run to ensure that every yaml file has a  last updated comment with a specific format so that the existing plugin by @Belleyy [GHScraperChecker](https://github.com/stashapp/CommunityScripts/tree/main/plugins/GHScraper_Checker) can check that out.  Any change in a scraper file (especially for scripts) requires(we double check that manually atm before merging the PR) that the commented date in the yml file is updated
>   This is probably a good starting point as the data is already in the yml file and could be parsed or propagated to the manifest or the index file

This sounds good. Will update the design.

>   Scraping operation should be run separately from the tasks. It is very common to do some scraping while waiting for scan/generation to finish. That would mean though that the identify task should be also moved? Allowing a single scrape at a  time will also  make it easier to limit/throttle requests during an identify task.

This is a misunderstanding of the design, and I will amend it. The operations mentioned in this design are scoped for scraper management tasks, _not scraping operations themselves_. Scraping operations are unchanged.

**Comment by scruffynerf** [comment]
Commenting to add myself to the discussion.  The above is also true for plugins, same principle.  I'm not sure I agree with the static page issue... And can think of a few counter arguments, including versioning and patch testing.  I'll flesh this out soonish.

Given the requirement of all scrapers and plugins having a .yml file, I'd be in favor of just adopting that file as the "main" file which lists other files part of the item, and potentially even versions of those, same with any pre-requirements (not necessarily python modules, which can be done with requirements.txt, and this process should be able to trigger off a pip request to the python system for that as well)
